app/models/models.scala

```diff
@@ -52,6 +52,9 @@ case class Instance(
   val amiArn = specification.get("imageArn")
 
   override def toString: String = s"Instance<$arn>"
+  def belongsToSSA(ssa: SSA): Boolean = ssa.stack == stack &&
```

Comment:
I try very hard to avoid mixing logic with data structure definitions with my own code. The only time I allow myself exceptions is when there is a property that is inherently a part of the object itself, which is to say it naturally arises as a part of the instance itself. It follows that this is only possible if it does't take any parameters.
One rule of thumb is that you shouldn't need any tests for the models package, which makes sense if it's just modelling data.

This is my taste so you're very welcome to disagree, but it looks to me like this would fit nicely with the rest of the logic?